src/ch9329.cpp

```diff
@@ -0,0 +1,401 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <array>
+#include <cstdint>
+#include <libhal-input/ch9329.hpp>
+#include <libhal-util/bit.hpp>
+#include <libhal-util/serial.hpp>
+#include <libhal/serial.hpp>
+#include <libhal/timeout.hpp>
+#include <libhal/units.hpp>
+
+namespace hal::input {
+
+constexpr hal::byte header_byte_1 = 0x57;
+constexpr hal::byte header_byte_2 = 0xAB;
+constexpr hal::byte address_byte = 0x00;
+
+constexpr hal::byte cmd_reset = 0x0F;
+constexpr hal::byte cmd_send_kb_general_data = 0x02;
+constexpr hal::byte cmd_send_kb_media_data = 0x03;
+constexpr hal::byte cmd_send_ms_abs_data = 0x04;
+constexpr hal::byte cmd_send_ms_rel_data = 0x05;
+
+constexpr auto header_frame_size = 5;
+
+ch9329::ch9329(hal::serial& p_uart)
+  : m_uart(&p_uart)
+{
+}
+
+ch9329::mouse_absolute::mouse_absolute(std::uint16_t p_screen_width,
+                                       std::uint16_t p_screen_height)
+  : m_screen_width(p_screen_width)
+  , m_screen_height(p_screen_height)
+
+{
+  m_data = { 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+}
+
+ch9329::mouse_relative::mouse_relative()
+{
+  m_data = { 0x01, 0x00, 0x00, 0x00, 0x00 };
+}
+
+ch9329::keyboard_media::keyboard_media()
+{
+  m_data = { 0x02, 0x00, 0x00, 0x00 };
+}
+
+ch9329::keyboard_acpi::keyboard_acpi()
+{
+  m_data = { 0x01, 0x00 };
+}
+
+ch9329::keyboard_general::keyboard_general()
+{
+  m_data = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
```

Comment:
This is just a comment, I didn't realize you could assign it like this. Very happy I used `hal::byte` and not `std::byte` because you'd need to wrap it in `std::byte{ 0x00 }` and so on. 🤮